include/ginkgo/core/distributed/matrix.hpp

```diff
@@ -467,20 +486,23 @@ class Matrix
      *                                   `create` returns. It should be the
      *                                   return value of make_matrix_template.
      */
-    template <typename LocalMatrixType, typename NonLocalMatrixType>
+    template <typename LocalMatrixType, typename NonLocalMatrixType,
+              typename = std::enable_if_t<
+                  detail::is_matrix_type_builder<LocalMatrixType, ValueType,
+                                                 LocalIndexType>::value &&
+                  detail::is_matrix_type_builder<NonLocalMatrixType, ValueType,
+                                                 LocalIndexType>::value>>
     explicit Matrix(std::shared_ptr<const Executor> exec,
                     mpi::communicator comm,
                     LocalMatrixType local_matrix_template,
                     NonLocalMatrixType non_local_matrix_template)
         : Matrix(exec, comm,
-                 static_cast<const LinOp*>(
-                     local_matrix_template
-                         .template create<ValueType, LocalIndexType>(exec)
-                         .get()),
-                 static_cast<const LinOp*>(
-                     non_local_matrix_template
-                         .template create<ValueType, LocalIndexType>(exec)
-                         .get()))
+                 local_matrix_template
+                     .template create<ValueType, LocalIndexType>(exec)
+                     .get(),
```

Comment:
Minor: I guess this means `Dense` cannot be a local/non-local type since it has no index type. It's probably not important considering we probably don't have a use case for that.

Reply:
It's still possible to create distributed matrix with local/non-local dense matrix type by using the `LinOp*` overloads. But in general I think a dense matrix wouldn't be distributed row-wise, so our approach doesn't make much sense there anyway.